centreon/src/Core/ResourceAccess/Domain/Model/DatasetFilter.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+/*
+ * Copyright 2005 - 2023 Centreon (https://www.centreon.com/)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Core\ResourceAccess\Domain\Model;
+
+use Assert\AssertionFailedException;
+use Centreon\Domain\Common\Assertion\Assertion;
+use Centreon\Domain\Common\Assertion\AssertionException;
+
+class DatasetFilter
+{
+    /** @var array<string, DatasetFilterType[]> */
+    private array $datasetFilterHierarchy = [];
+
+    private ?self $datasetFilter = null;
+
+    /**
+     * @param string $type
+     * @param int[] $resourceIds
+     *
+     * @throws AssertionFailedException
+     * @throws AssertionException
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(
+        private readonly string $type,
+        private readonly array $resourceIds,
+    ) {
+        $shortName = (new \ReflectionClass($this))->getShortName();
+        Assertion::notEmptyString($type, "{$shortName}::type");
+
+        // try to convert string type to Enum to validate that type is supported
+        DatasetFilterTypeConverter::fromString($type);
+
+        Assertion::notEmpty($this->resourceIds, "{$shortName}::resourceIds");
+        Assertion::arrayOfTypeOrNull('int', $this->resourceIds, "{$shortName}::resourceIds");
+        $this->createDatasetFilterHierarchy();
+    }
+
+    /**
+     * @return string
+     */
+    public function getType(): string
+    {
+        return $this->type;
+    }
+
+    /**
+     * @return int[]
+     */
+    public function getResourceIds(): array
+    {
+        return $this->resourceIds;
+    }
+
+    /**
+     * @param DatasetFilter|null $datasetFilter
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function setDatasetFilter(?self $datasetFilter): void
+    {
+        if ($datasetFilter !== null) {
+            $this->assertHierarchyIsValid($datasetFilter);
+        }
+        $this->datasetFilter = $datasetFilter;
+    }
+
+    /**
+     * @return null|DatasetFilter
+     */
+    public function getDatasetFilter(): ?self
+    {
+        return $this->datasetFilter;
+    }
+
+    private function createDatasetFilterHierarchy(): void
```

Comment:
I would rather use a dedicated object for this structure : 
- split the logic
- allow testing the structure logic (ie. avoiding cycling changes etc ...)

Reply:
That will be done later on